### ISSUE TITLE
Round-tripping wipes out a considerable amount of data

### ISSUE THREAD

**Issue by DanielHeath** [radiopaedia/dicomanon#8]
After the first run the file is just under 80kb; putting it through the example again drops that to 59kb, and a third time drops it to 38k (at which point it's no longer parsable as a dicom file according to dicomanon)

[dicoms.zip](https://github.com/radiopaedia/dicomanon/files/5121777/dicoms.zip)

**Comment by kivikakk** [comment]
I note that OsiriX shows the twice/thrice-throughs as plain gradients, possibly just what it shows if it can't find any image data.

<img width="663" alt="image" src="https://user-images.githubusercontent.com/1915/91139750-c8967880-e6f2-11ea-97ff-3122264a77d0.png">

**Comment by plwp** [comment]
Cool bug...

My best guess is that there's a value length that's being set wrong in one or more of the tags and that's causing either a read or write to be offset? 

I can't see anything glaring but would be worth trying to work out if the writer is writing the same amount of data as the value length it's calculating.
https://github.com/radiopaedia/dicomanon/blob/56c3a97fb73dbe1641dbc97d3818b61d7e3e5ccf/src/Tag.ts#L86

I'll have a poke around with a hex editor when I get a chance to see if anything looks obviously malformed...

**Comment by plwp** [comment]
Another possibility is that something's funky's happening with the whitespace trimming on read for some of the text tags and that's somehow breaking the format on write?

https://github.com/radiopaedia/dicomanon/blob/56c3a97fb73dbe1641dbc97d3818b61d7e3e5ccf/src/ValueRepresentation.ts#L8

**Comment by DanielHeath** [comment]
Tried `rtrim` - for the files I'm getting errors on, it didn't change *anything* in the output.

**Comment by DanielHeath** [comment]
A quick test confirms it's definitely not related to anonymizing

**Comment by plwp** [comment]
Looks like it's eating exactly 20488 (0x5008) bytes starting at 043B in a continuous chunk. 

Having a look at Message.ts to see if there's any likely culprits...

**Comment by DanielHeath** [comment]
I'm moderately suspicious of the resizing code in the custom buffer implementation - that would be a plausible mechanism to cut out a bunch of data at a fixed offset

**Comment by DanielHeath** [comment]
Got a failing test case confirming it's in the custom buffers

**Comment by plwp** [comment]
Very TDD of you :) Nice work!

I would have just stuck a bunch of print statements in and around the concat method to see what was getting called just before and after the offset. :P

Sounds like you have it sorted though.

**Comment by DanielHeath** [comment]
I deleted that after a second, realizing that I need to confirm that the bug isn't in the test

**Comment by DanielHeath** [comment]
Also, I'm not getting 043B as the offset - with a pure round-trip (no anon), I get the first diff at 309D.

**Comment by DanielHeath** [comment]
I feel like `readHex` should error out if it reads data for which `this.readUint8().toString(16);` doesn't give you a single character.

**Comment by DanielHeath** [comment]
Ahh. `readHex` and `writeHex` are not opposites (seems if you pass a 40-character string to writeHex and it'll write 20 bytes; read 20 bytes with readHex and it'll return a 20 character string).

**Comment by kivikakk** [comment]
readHex should deeeefinitely be padding digits, that's super broken.

**Comment by DanielHeath** [comment]
This is two layers lower in the stack than I usually work; I'm still making progress, but it involves dusting off some long-dormant skills.

To confirm:

```
const {WriteBufferStream} = require('../lib/BufferStream')
var fileStream = new WriteBufferStream(4096, true);
fileStream.writeHex("000102030405060708090A00000000000000000000000000000000000000000")
fileStream.offset = 0
console.log(fileStream.readHex('000102030405060708090A'.length)) // 0123456789a00000000000
```
I would expect that console.log to return the original string.

**Comment by kivikakk** [comment]
Yep — it should look approximately like this:

```typescript
    readHex(length) {
        var hexString = '';
        for (var i = 0; i < length; i++) {
            hexString += this.readUint8().toString(16).padStart(2, '0');
        }
        return hexString;
    }
```

**Comment by plwp** [comment]
Full stack means field programming a gate array in CSS.

**Comment by DanielHeath** [comment]
I designed a CPU once, but that was 15 years ago; the last 5 years I've mostly dealt with "oh, you meant *that* interpretation of that word"

**Comment by kivikakk** [comment]
I once worked on a dynamic assembly JIT for CSS selectors. Maybe the best/worst day of my programming career.

**Comment by DanielHeath** [comment]
Wait, if `padStart` exists how did the left-pad fiasco happen?

**Comment by kivikakk** [comment]
<img width="627" alt="image" src="https://user-images.githubusercontent.com/1915/91388585-0a522b00-e87b-11ea-948b-84e6ae956701.png">

<img width="470" alt="image" src="https://user-images.githubusercontent.com/1915/91388573-032b1d00-e87b-11ea-9390-0c65b2a7a178.png">

🥁

**Comment by DanielHeath** [comment]
Depressingly, half our uploads come from ie11 machines (presumably in hospitals; I don't imagine many people install it at home on purpose)

**Comment by DanielHeath** [comment]
Alright, fixing the padding doesn't actually fix the truncation, so I'll keep looking

**Comment by DanielHeath** [comment]
So, I've instrumented `WriteBufferStream` so that all writes to the 'root' buffer stop when any written byte doesn't match the original, then run a file through a roundtrip (load / dump).

The first place it stops is in `concat`, when adding the pixeldata tag (with a `BinaryRepresentation`). That exactly matches the 309D offset I was seeing earlier!

So, it looks like somehow the pixel data isn't getting transcribed correctly. Signing off for the week - hopefully I'll be able to figure it out early next week.

**Comment by plwp** [comment]
So it sounds like it tries to write the pixel data into the buffer, expands the buffer till it gets to 0x5000 large, writes the next byte, drops the buffer, thinks its doing a bang-up job, then for some reason never has that problem again...

**Comment by DanielHeath** [comment]
None of the other tags get resized (but then, none are close to the same size)

**Comment by DanielHeath** [comment]
I can't help but notice that `var fragmentSize = 1024 * 20` in the tag-writing code is the exact amount of data getting chopped off...

**Comment by DanielHeath** [comment]
Wait a second - could the code in `BinaryRepresentation` be assuming that the transfer syntax at write-time `EXPLICIT_LITTLE_ENDIAN`? 

I've got a file with `meta['00020010'].Value == '1.2.840.10008.1.2.4.70'`. Could that make the offsets for binary data get encoded wrong?

**Comment by DanielHeath** [comment]
Ah crap. I've been looking at the writer all this time, and feel terribly silly for realizing that it's fine, but the data isn't being *read* correctly

**Comment by plwp** [comment]
To be fair, it FEELS like a write problem...

**Comment by DanielHeath** [comment]
Sigh. I'm all over the place; now I switch to another file and the read seems fine but the write is off. Perhaps this isn't the ideal place for a stream-of-consciousness debug log...

**Comment by plwp** [comment]
I've tried to step through read/writing a file and the fragment size definitely looks suspicious, but I can't see it dropping any. It's definitely still killing the data somewhere.

If it's happening on read (sometimes) and write (sometimes) you have to be suspicious of the shared code in BufferStream. I'm looking at you `more` and `concat`.

**Comment by plwp** [comment]
So there was an issue in checkSize where it wouldn't expand longer than 2x the current size. Fixed it but it wasn't the problem.
Fixed the hex padding with an IE friendly method, obviously that wasn't it.

The pixel stream gets read fine, but when it's getting the converted to a tag value is where it's losing the frame.
I've narrowed it down to the ValueRepresentation.readBytes function, but it's looking like it'll be a real pain to debug without going into the DICOM spec and working out how everything's packed. 

Honestly this library is buggy enough that it's worth spending that effort on shifting to the dicom-parser library.
